irohad/wsv_checker/wsv_checker.cpp

```diff
@@ -0,0 +1,1119 @@
+
+template<typename _CharT, typename _Traits, typename _Alloc>
+inline void
+unquote(std::basic_string<_CharT, _Traits, _Alloc>& data)
+{
+   _CharT const* rptr = data.data();
+   _CharT* wptr = (_CharT*)rptr;
+   if (rptr[0] != _CharT('\0'))
+      do {
+         rptr += *rptr == _CharT('\\') ? 1ul : 0ul;
+         *wptr++ = *rptr;
+      } while (*rptr++ != _CharT('\0'));
+   data.resize(data.size() - size_t(rptr - wptr));
+}
+
+struct Peer
+{
+   std::string pubkey;
+   std::string address;
+   mutable std::string tls;
+
+   bool operator<(Peer const& p) const { return pubkey < p.pubkey; }
+   friend ostream& operator<<(ostream& os, Peer const& p)
+   {
+      return os << "\n  "
+                << "pubkey:" << p.pubkey << " address:" << p.address << " tls:" << p.tls;
+   }
+   static Peer from_soci_row(soci::row const& row)
+   {
+      Peer p;
+      row >> p.pubkey >> p.address;
+      if (row.get_indicator(2) != soci::i_null)
+         row >> p.tls;
+      return p;
+   }
+   bool check_equals(Peer const& o) const
+   {
+#if 0
+      if (pubkey != o.pubkey) {
+            fmt::print("Peers have different pubkey: '{}' and '{}'\n", pubkey, o.pubkey);
+            return false;
+         }
+         if (address != o.address) {
+            fmt::print("Peers have different address: '{}' and '{}'\n", address, o.address);
+            return false;
+         }
+         if (tls != o.tls) {
+            fmt::print("Peers have different tls: '{}' and '{}'\n", tls, o.tls);
+            return false;
+         }
+#endif
+      CHECK_EQUALS_RETURN("Peers", pubkey, o.pubkey)
+      CHECK_EQUALS_RETURN("Peers", address, o.address)
+      CHECK_EQUALS_RETURN("Peers", tls, o.tls)
+      return true;
+   }
+};
+struct Role
+{
+   std::string name;
+   std::string permissions;
+
+   bool operator<(Role const& o) const { return name < o.name; }
+   friend ostream& operator<<(ostream& os, Role const& r)
+   {
+      return os << "\n  "
+                << "name:" << r.name << " permisions:" << r.permissions;
+   }
+   static Role from_soci_row(soci::row const& r)
+   {
+      Role o;
+      r >> o.name >> o.permissions;
+      return o;
+   }
+   static Role from_key_value(string_view& key, string_view val)
+   {
+      auto r = Role{std::string{key.substr(0, key.find('/'))}, std::string(val)};
+      key = key.substr(r.name.size() + 1);   // bypass '/'
+      return r;
+   }
+   bool check_equals(Role const& o) const
+   {
+      // CHECK_EQUALS("Roles", name, o.name)
+      // CHECK_EQUALS("Roles", permissions, o.permissions)
+      if (name != o.name) {
+         fmt::print("Roles have different name: '{}' and '{}'\n", name, o.name);
+         return false;
+      }
+      if (permissions != o.permissions) {
+         fmt::print("Roles have different permissions: '{}' and '{}'\n", permissions, o.permissions);
+         return false;
+      }
+      return true;
+   }
+};
+struct AssetPrecision
+{
+   std::string name;
+   mutable long double precision;
```

Comment:
Why ```mutable```?

Reply:
because   std::set<AssetPrecision> where the key is name.